crypto/ex_data.c

```diff
@@ -32,15 +33,42 @@ typedef struct ex_callbacks_st {
 
 static EX_CALLBACKS ex_data[CRYPTO_EX_INDEX__COUNT];
 
+/*
+ * There is the possibility that initialization happens more than once,
+ * for example with provider modules being initialized multiple times,
+ * but also with multiple library contexts.
+ * However, the inner EX_DATA stuff is global for the shared library or
+ * module, and should therefore only be initialized once (on first use)
+ * and cleaned away once (on last cleanup) in the life time of the shlib
+ * or module.
+ *
+ * The most practical way to handle this is with a reference count.
+ */
+static CRYPTO_REF_COUNT ex_data_lock_count = 0;
+static CRYPTO_RWLOCK *ex_data_lock_count_lock = NULL;
+
 static CRYPTO_RWLOCK *ex_data_lock = NULL;
-static CRYPTO_ONCE ex_data_init = CRYPTO_ONCE_STATIC_INIT;
+
+static CRYPTO_ONCE ex_data_init_flag = CRYPTO_ONCE_STATIC_INIT;
 
 DEFINE_RUN_ONCE_STATIC(do_ex_data_init)
 {
-    if (!OPENSSL_init_crypto(0, NULL))
+    if (!OPENSSL_init_crypto(0, NULL)
+        || (ex_data_lock_count_lock = CRYPTO_THREAD_lock_new()) == NULL
+        || (ex_data_lock = CRYPTO_THREAD_lock_new()) == NULL) {
+        CRYPTO_THREAD_lock_free(ex_data_lock);
+        CRYPTO_THREAD_lock_free(ex_data_lock_count_lock);
         return 0;
-    ex_data_lock = CRYPTO_THREAD_lock_new();
-    return ex_data_lock != NULL;
+    }
+    return 1;
+}
+
+static int ex_data_init(void)
+{
+    int i;
+
+    return RUN_ONCE(&ex_data_init_flag, do_ex_data_init)
+        && CRYPTO_UP_REF(&ex_data_lock_count, &i, ex_data_lock_count_lock);
```

Comment:
This doesn't look right. We'll end up incrementing the ref count every time someone calls `get_and_lock`.

Reply:
Hmmm, you're right.  Damnit!

This is quite infuriating, it seems that we're fighting a losing battle to avoid having to deal with shared object init/fini, DllMain or whatever different platforms use, while trying to perform exactly what they are designed for.

Reply:
Hey....did I ever mention openssl_ctx_runone()... ;-)